driver/driver.go

```diff
@@ -384,38 +384,51 @@ func (c *Conn) Prepare(query string) (driver.Stmt, error) {
 
 // ExecContext is an optional interface that may be implemented by a Conn.
 func (c *Conn) ExecContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Result, error) {
-	ctx, span := tracing.Start(ctx, "dqlite.driver.ExecContext", query)
-	defer span.End()
-
-	if int64(len(args)) > math.MaxUint32 {
-		return nil, driverError(c.log, fmt.Errorf("too many parameters (%d)", len(args)))
-	} else if len(args) > math.MaxUint8 {
-		protocol.EncodeExecSQLV1(&c.request, uint64(c.id), query, args)
-	} else {
-		protocol.EncodeExecSQLV0(&c.request, uint64(c.id), query, args)
-	}
-
-	var start time.Time
-	if c.tracing != client.LogNone {
-		start = time.Now()
-	}
-	err := c.protocol.Call(ctx, &c.request, &c.response)
-	if c.tracing != client.LogNone {
-		c.log(c.tracing, "%.3fs request exec: %q", time.Since(start).Seconds(), query)
-	}
-	if err != nil {
-		return nil, driverError(c.log, err)
-	}
+    ctx, span := tracing.Start(ctx, "dqlite.driver.ExecContext", query)
+    defer span.End()
+
+    var result protocol.Result
+	// var error error
+    var err error
+
+    for {
+        if int64(len(args)) > math.MaxUint32 {
+            return nil, driverError(c.log, fmt.Errorf("too many parameters (%d)", len(args)))
+        } else if len(args) > math.MaxUint8 {
+            protocol.EncodeExecSQLV1(&c.request, uint64(c.id), query, args)
+        } else {
+            protocol.EncodeExecSQLV0(&c.request, uint64(c.id), query, args)
+        }
+
+        var start time.Time
+        if c.tracing != client.LogNone {
+            start = time.Now()
+        }
+        err = c.protocol.Call(ctx, &c.request, &c.response)
+        if c.tracing != client.LogNone {
+            c.log(c.tracing, "%.3fs request exec: %q", time.Since(start).Seconds(), query)
+        }
+		if err != nil {
+			return nil, driverError(c.log, err)
+		}
+        result, err = protocol.DecodeResult(&c.response)
+        if err != nil && c.tracing != client.LogNone {
 
-	var result protocol.Result
-	result, err = protocol.DecodeResult(&c.response)
-	if err != nil {
-		return nil, driverError(c.log, err)
-	}
+			duration := 20 * time.Microsecond
+			time.Sleep(duration)
+            c.log(c.tracing, "error in exec context %v", err)
```

Comment:
@SaumyaSachdev question are you attempting to retry on failure?

Currently, the likes of LXD and Juju are retrying at a higher level (see LXD for an example):

https://github.com/canonical/lxd/blob/d5b1d00f5651a6b46541e6787d4f9a6ed494f179/lxd/db/query/retry.go#L25-L58

I would like to move this into the driver, but it should be configurable if it was placed here.


Note: the title to the PR is incorrect for what is happening.